MobileNetV2 keras implementation for imagenet weights for Tensorflow backend.

JonathanCMitchell opened the pull request:
This file enables access to the 23 trained MobileNetV2 models found [here](https://github.com/tensorflow/models/tree/master/research/slim/nets/mobilenet). The design and architecture for MobileNetV2 is explained in this [paper](https://arxiv.org/abs/1801.04381)

For a complete view of the weight extraction. loading, and testing process please see this [repository](https://github.com/JonathanCMitchell/mobilenet_v2_keras)

Link to google doc explaining changes:
https://docs.google.com/document/d/1-sRVeWiUMNmM8uZSfX-V0dZu9D_580OOQaTKdoqcJHM/edit?usp=sharing

Please let me know if there is something you would like me to add/ change/ etc.

taehoonlee commented:
Thank you for the PR @JonathanCMitchell. Please fix PEP8 [errors](https://travis-ci.org/keras-team/keras/jobs/371840019) and add a unit test.

JonathanCMitchell commented:
ok I figured out how to run the tests, but I am failing tests for files I did not touch. Such as /tests/keras/backend/backend_test.py for 

```
[gw0] [  3%] FAILED keras/tests/test_model_saving.py::test_loading_weights_by_name_skip_mismatch 
keras/tests/test_model_saving.py::test_saving_lambda_numpy_array_arguments 

```

```
[gw1] [ 20%] FAILED keras/tests/keras/backend/backend_test.py::TestBackend::test_batchnorm 
keras/tests/keras/backend/backend_test.py::TestBackend::test_ctc 
```
I am pretty clueless on what to do. I can try to mimic other PR's testing schemes but it would help to have some guidance. Thanks!

taehoonlee commented:
@JonathanCMitchell, You don't need to worry about the two failed tests (`test_loading_weights_by_name_skip_mismatch` and `test_batchnorm`). They have nothing to do with your work. We would be grateful if you could focus on the following two items.

- Fix PEP8 [errors](https://travis-ci.org/keras-team/keras/jobs/372267775#L1184),
- Update the number of classes: Your model seems to have 1001 classes including a background class (see the [travis log](https://travis-ci.org/keras-team/keras/jobs/372267778#L2281)). Can you exclude the background class for both the model definition and the twenty six weight files?

Thank you for your great work!

JonathanCMitchell commented:
Hi @taehoonlee I just added the following fixes:
(1) I fixed the pep8 issues. 
(2) I changed the classes from 1001 -> 1000 and updated all the weight files to meet this requirement.
(3) I added support to infer the shape of the input tensor from either `input_tensor` or `input_shape`, and to check discrepancies if both are provided. I also added some error handling for this feature. This should be pretty helpful in the future.

I also passed all the tests (including pep8), except for the same batch_norm one that I shouldn't worry about.

Please let me know if there is anything further that I can do.

Best wishes,
Jonathan

taehoonlee commented:
There is something wrong. I tested your `MobileNetV2` with the following codes:

```python
from keras.applications.mobilenetv2 import MobileNetV2
from keras.preprocessing import image
from keras.applications.mobilenetv2 import preprocess_input, decode_predictions
import numpy as np

model = MobileNetV2(weights='imagenet')

img_path = '../cat.png'
img = image.load_img(img_path, target_size=(224, 224))
x = image.img_to_array(img)
x = np.expand_dims(x, axis=0)
x = preprocess_input(x)

preds = model.predict(x)
# decode the results into a list of tuples (class, description, probability)
# (one such list for each sample in the batch)
print('Predicted:', decode_predictions(preds, top=3)[0])
```

The results are `Predicted: [(u'n03000247', u'chain_mail', 0.38690498), (u'n02018795', u'bustard', 0.027022662), (u'n03482405', u'hamper', 0.024686841)]`.

There must be some mistake during slicing weight matrices of the last fully connected layers from `(?, 1001)` to `(?, 1000)`.

taehoonlee commented:
@JonathanCMitchell, Please note that I tested with TensorFlow==1.3.0 and 1.4.0.

JonathanCMitchell commented:
@taehoonlee ok I fixed it. I forgot to add a numpy dependency because the preprocessing function casts the output to `np.float32`. 

In your case, you ran the network previously (before the most recent changes), so you have the previous model stored in your cache. Therefore, if you remove this file, you will download a new one. Please go to 

You have to delete the model stored in your cache, so that the new model can be downloaded / updated. Check `/.keras/models` and delete the mobillenet2 model that lives there.

my output: (I used a panda picture: 
```
pe`.
Using TensorFlow backend.
Predicted: [('n02510455', 'giant_panda', 0.85812426), ('n04254680', 'soccer_ball', 0.009869265), ('n03761084', 'microwave', 0.0026237909)]
```

fchollet commented:
Thanks for the PR, this is useful addition, so we should go with it. I'll leave it to @taehoonlee for the review. Thanks!

taehoonlee commented on keras/applications/mobilenetv2.py:
```diff
@@ -0,0 +1,535 @@
+"""MobileNet v2 models for Keras.
+
+MobileNetv2 is a general architecture and can be used for multiple use cases.
```
Would it be better to make these words (`MobileNet v2`, `MobileNetv2`, `MobileNetV2`) consistent?

taehoonlee commented on tests/keras/applications/applications_test.py:
```diff
@@ -144,6 +144,16 @@ def test_mobilenet():
     _test_app_pooling(app, last_dim)
 
 
+@pytest.mark.skipif((K.backend() != 'tensorflow'),
```
Please enable the test for all the backends.

taehoonlee commented on keras/applications/mobilenetv2.py:
```diff
@@ -0,0 +1,535 @@
+               kernel_size=3,
+               strides=(2, 2), padding='same',
+               use_bias=False, name='Conv1')(img_input)
+    x = BatchNormalization(epsilon=1e-5, name='bn_Conv1')(x)
```
For enabling on Theano, please modify `epsilon=1.01e-5` (simple trick).

JonathanCMitchell replied:
Do I need to do this with the epsilon value of 1e-3?

taehoonlee commented on keras/applications/mobilenetv2.py:
```diff
@@ -0,0 +1,535 @@
+
+The weights for all 16 models are obtained and translated from the Tensorflow checkpoints
+from TensorFlow checkpoints found at
+https://github.com/tensorflow/models/blob/master/research/slim/nets/mobilenet_v1.md
```
These lines must be revised with respect to v2 or removed.

taehoonlee commented on keras/applications/mobilenetv2.py:
```diff
@@ -0,0 +1,535 @@
+(https://github.com/JonathanCMitchell/mobilenet_v2_keras)
+
+Tests comparing this model to the existing Tensorflow model can be
+found at [mobilenet_v2_keras])(https://github.com/JonathanCMitchell/mobilenet_v2_keras)
```
Typo.

taehoonlee replied:
`)` after `]`

taehoonlee commented on keras/applications/mobilenetv2.py:
```diff
@@ -0,0 +1,535 @@
+[MobileNetV2: Inverted Residuals and Linear Bottlenecks](https://arxiv.org/abs/1801.04381)
+
+The weights for this model were extracted, loaded, and tested using this repository:
+(https://github.com/JonathanCMitchell/mobilenet_v2_keras)
```
Redundant link or description.

taehoonlee commented on keras/applications/mobilenetv2.py:
```diff
@@ -0,0 +1,535 @@
+                input_tensor=None,
+                classes=1000):
+    """
+    Instantiates the MobileNetv2 architecture.
```
Please remove new line (`"""Instantiates`).

taehoonlee commented on keras/applications/mobilenetv2.py:
```diff
@@ -0,0 +1,535 @@
+    # Raises
+        ValueError: in case of invalid argument for `weights`,
+            or invalid input shape.
+        RuntimeError: If attempting to run this model with a
```
Please remove.

taehoonlee replied:
I mean `RuntimeError`.

taehoonlee commented on keras/applications/mobilenetv2.py:
```diff
@@ -0,0 +1,535 @@
+
+    # Raises
+        ValueError: in case of invalid argument for `weights`,
+            or invalid input shape.
```
or invalid depth_multiplier, alpha, rows when `weights='imagenet'`.

taehoonlee commented on keras/applications/mobilenetv2.py:
```diff
@@ -0,0 +1,535 @@
+
+    x = Conv2D(expansion * in_channels, kernel_size=1, padding='same',
+               use_bias=False, activation=None,
+               name='mobl%d_conv_%d_expand' % (block_id, block_id))(inputs)
```
The `mobl%d` looks unnecessary.

taehoonlee commented on keras/applications/mobilenetv2.py:
```diff
@@ -0,0 +1,535 @@
+    x = Conv2D(expansion * in_channels, kernel_size=1, padding='same',
+               use_bias=False, activation=None,
+               name='mobl%d_conv_%d_expand' % (block_id, block_id))(inputs)
+    x = BatchNormalization(epsilon=1e-5, name='bn%d_conv_%d_bn_expand' %
```
The `bn%d` looks unnecessary.

taehoonlee commented on keras/applications/mobilenetv2.py:
```diff
@@ -0,0 +1,535 @@
+    # Depthwise
+    x = DepthwiseConv2D(kernel_size=3, strides=stride, activation=None,
+                        use_bias=False, padding='same',
+                        name='mobl%d_conv_%d_depthwise' %
```
Same.

taehoonlee commented on keras/applications/mobilenetv2.py:
```diff
@@ -0,0 +1,535 @@
+    # Project
+    x = Conv2D(pointwise_filters,
+               kernel_size=1, padding='same', use_bias=False, activation=None,
+               name='mobl%d_conv_%d_project' % (block_id, block_id))(x)
```
Same.

taehoonlee commented on keras/applications/mobilenetv2.py:
```diff
@@ -0,0 +1,535 @@
+    x = DepthwiseConv2D(kernel_size=3,
+                        strides=stride, activation=None,
+                        use_bias=False, padding='same',
+                        name='mobl%d_conv_%d_depthwise' %
```
The same.

taehoonlee commented on keras/applications/mobilenetv2.py:
```diff
@@ -0,0 +1,535 @@
+               padding='same',
+               use_bias=False,
+               activation=None,
+               name='mobl%d_conv_%d_project' %
```
The same.

taehoonlee commented:
The codes look good but there is still something wrong for prediction. The reported top-1 and top-5 accuracies are 71.8% and 91.0%, respectively. However in my tests, the performances were 58.558% and 81.408%.

This issue is not from `Conv2D(strides=2, padding='same')`. Maybe from BatchNormalization or DepthwiseConv2D? We need to take a closer look.

JonathanCMitchell commented:
@taehoonlee I will fix all the comments/naming stuff but we should probably handle the performance issue first. I re-confirmed all the weights that were extracted have been loaded. Specifically. 
```
BatchNorm layers: 52
Conv2D layers: 35
DepthwiseConv2D layers: 17
Activations: 35 + 1 (softmax built into last Dense layer)
Dense: 1
Flatten: 1
AveragePool2D: 1
Input: 1
Add: 10 (if you are using alpha = 1.0, resolution = 224. )
``` 
For each BatchNorm layer: there are 4 parameters. [gamma, beta, moving_mean, moving_variance] that are loaded in that order. When I inspect the model after:
```
model = MobileNetV2(weight='imagenet')
```
I notice that `gamma` and `beta` are still set to trainable. 

I noticed after searching around that tensorflow models come preloaded with a background class that is fine-tuned 

I am not sure if the tensorflow models come preloaded with a background class that has been tuned. If this is the case, their use of 1001 classes encompasses that background class. If that bg class has been fine-tuned on images that do not contain any classes, then they have an additional weight value in their FC layer that stores that information, and could be the reason why our results are not as good as theirs. 

In my inference tests, which can be found [here](https://github.com/JonathanCMitchell/mobilenet_v2_keras/blob/master/test_mobilenet.py#L100), I simply popped out the first layer and then re-normalized the logit values, removing that background class.

Tensorflow data for last layer:
```
 {'block_id': 19,
  'key': 'MobilenetV2/Logits/Conv2d_1c_1x1/weights',
  'layer': 'Logits',
  'meta': 'weights',
  'mod': '',
  'shape': (1, 1, 1280, 1001)},
 {'block_id': 19,
  'key': 'MobilenetV2/Logits/Conv2d_1c_1x1/biases',
  'layer': 'Logits',
  'meta': 'biases',
  'mod': '',
  'shape': (1001,)}]
```
So when I loaded the weights I used `weights[:, 1:]` and `biases[1:]`. This can be seen [here](https://github.com/JonathanCMitchell/mobilenet_v2_keras/blob/3c8ce250baff6b07879c68b4e3e02ab08e82a5e1/load_weights_multiple.py#L194)

Do you happen to know what class fails the test? I do not have the imagenet dataset saved locally and I currently don't have the resources to run the evaluation.

JonathanCMitchell commented:
I grabbed the first feature map after their first Conv2D + BatchNorm + Relu and checked the feature map against mine, and we have different values. I checked the weights in that `Conv2D` and `BatchNorm` layers and we have the same exact values and shape. Therefore, the way their convolutional kernels operate is different from ours. I've concluded that all the weights have been loaded properly.

It might have something to do with [_fixed_padding](https://github.com/tensorflow/models/blob/master/research/slim/nets/mobilenet/conv_blocks.py#L24), but I have yet to figure out how to make ours match.

I tried grabbing the output after the first convolutional operation, but I could not find a pointer to it from the checkpoint file alone.

titu1994 commented:
For the additional class, simply splicing out the first index of the final weight and bias should be sufficient. 

For the weights that you ported, you can check against the port in my repo (https://github.com/titu1994/MobileNetworks).

The model for V1 and V2 are separated inside the singular mobilenet.py. Inferece on random cheetah and elephant images seems to give correct class values. 

For the weight transfer, mine is a bit more hands-on, and can be seen in _weight_transfer_v2 folder. For the background check removal, it's the last few lines inside the weight transfer scripts there.

JonathanCMitchell commented:
@titu1994 Did your values meet the same specs as the TF models?

titu1994 commented:
@JonathanCMitchell By values do you mean the number of parameters and the floating point operations? I only checked for the alpha = 1.0 models, but for those, # of traininable parameters matches the paper, and FLOPs is approximately double that of the paper (when halved, it matches the papers with a small difference).

JonathanCMitchell commented:
@titu1994 I mean the imagenet prediction score or the feature map outputs. I am in the process of running your model to extract the first feature map and compare it against the tensorflow one for the same input.

JonathanCMitchell commented:
@titu1994 Hmm that's interesting that the _make_divisible didn't break the weight loading because I didn't use it. Tomorrow in the AM I will compare the actual weight values of both of our models and the outputs from the first feature map, (after the first Conv + BN + Relu).

When you said `EWA` weights did you mean the weights from the frozen model? I was benchmarking against the frozen model and tomorrow I can benchmark against the ordinary model.

titu1994 commented:
@JonathanCMitchell Yes I was looking at the EWA weights, and erroneously made the assumption that the session would load the EWA weights of the model. I don't think that is the case though.

JonathanCMitchell commented:
@titu1994 Please tell me what `EWA` stands for / what it means?

JonathanCMitchell commented:
I can tell you that I ignored all the RMS_Prop stuff, and basically anything that didn't have to do with my model / weights. I described the layers that I loaded in an above comment.

titu1994 commented:
EWA = Exponential Weighted Average. AKA Exponential Moving Averages. When I extract the weights, there are two sets of each weights. One is the original, and one is called the `_moving_average` weights. Therefore there are two sets of most of the weights (not all of them for some reason)

The reason they are used extensively in TF, is that they generally have better final performance on the test set as compared to the actual trained model with the weights at the final training step. I usually delete these exponential weighted average weights, as it makes the task of conversion comparatively harder.

I am wondering if we instead use the EWA weights in the correct places, will it perhaps fix the reduced prediction confidence that we are observing in our models.

JonathanCMitchell commented:
Gotcha! so I loaded the checkpoint with `tf.train.NewCheckpointReader` and then read all the stuff in the checkpoint file, and used `get_tensor(<tensorname>)` to receive them. I am probably not using the EWA weights. see [here](https://github.com/JonathanCMitchell/mobilenet_v2_keras/blob/master/extract_weights.py#L26)

titu1994 commented:
From this [link that you posted for mobilenet evaluation](https://github.com/tensorflow/models/blob/master/research/slim/nets/mobilenet/mobilenet_example.ipynb)

> Restore using exponential moving average since it produces (1.5-2%) higher 
> accuracy
```python
ema = tf.train.ExponentialMovingAverage(0.999)
vars = ema.variables_to_restore()

saver = tf.train.Saver(vars)
```

I doubt the NewCheckpointReader does this step automatically

JonathanCMitchell commented:
@titu1994 That makes sense because the weights I have are probably tuned from the last image sent through instead of the EMA of all the weights.

titu1994 commented:
@JonathanCMitchell I just tried with EWA weights ported for Alpha = 1 and 224 image size. Locally, the scores improved significantly. I suggest trying out at least one alpha x image size combination and seeing if it matches the performance stated by Tensorflow in that evaluation script.

JonathanCMitchell commented:
@titu1994 I'm not on my linux machine so I can't test atm, but how did you extract the EWA weights? Did you use tf.train.NewCheckpointReader? If I can do it for one of them, I will do it for all the weight files

titu1994 commented:
@JonathanCMitchell My procedure is different from yours by a lot. I am extracting the weights at the numpy level into separate files and then loading them up layer by layer. [See this directory for weight extraction and loading](https://github.com/titu1994/MobileNetworks/tree/master/_weight_extraction_v2).

I would think that NewCheckpointReader does get the EMA weights, however, it does not have a reference to the `tf.train.ExponentialMovingAverages(0.999)` part, so it is unable to restore it ? I don't use EMA much so I don't know the exact procedure.

Blindly following the evaluation script, the steps would be something akin to : 

1) Build MobileNet graph
2) Create the EMA object and get its variables
3) Tell Saver to restore only those variables and then use a Session to restore the graph with its original weights + EMA weights. 

However, at this step, how one would go about extracting just the EMA weights, I am unable to answer. 

As much as I can tell, my scripts to extract numpy weights and then load them up by hand seem to work correctly. Since our models match in parameter size, you should be able to drop-in replace my model with yours and use my weight extraction + loading script without any change.

JonathanCMitchell commented:
@titu1994 Yeah I do pretty much the same thing as you, I actually realized that I was just ignoring the EMA weights entirely and using the normal weights instead. In my approach I save the numpy files, but I also generate a guide that records the shape, metadata, and other information about the layers so I can be certain that they fill the right slot. Did you update the model file in your git release so I can try to plug N play with it tomorrow?

titu1994 commented:
I will push the updated weight file for MobileNetV2 alpha=1.0 ; 224x224 image size to my repo tonight. You will need to delete the older weight file if you have installed it inside ~/.keras/models/

JonathanCMitchell commented:
@titu1994 will do! Thank you!! I'll let you know tomorrow what the results are.

taehoonlee commented:
@JonathanCMitchell, @titu1994:
There is slight discrepancy between the @JonathanCMitchell's weight files and the original repo; but it is the behaviour of `BatchNormalization` that has more impact on the difference results.

My test codes are:
```python
import numpy as np
import tensorflow as tf
import tensornets as nets
import tensorflow_hub as hub

img = nets.utils.load_img('cat.png', target_size=256, crop_size=224)
img = (img / 128.0) - 1.0

inputs = tf.placeholder(tf.float32, [None, 224, 224, 3])
model = hub.Module("https://tfhub.dev/google/imagenet/mobilenet_v2_100_224/classification/1")
features = model(inputs, signature="image_classification", as_dict=True)
probs = tf.nn.softmax(features['default'])

with tf.variable_scope('keras'):
    model2 = MobileNetV2(weights='imagenet')

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    preds = sess.run(probs, {inputs: img})

preds2 = model2.predict(img)

print(nets.utils.decode_predictions(preds[:, 1:]))
print(nets.utils.decode_predictions(preds2))
```

The results are:
- The reference: `[[(u'n02124075', u'Egyptian_cat', 0.18302459), (u'n02123045', u'tabby', 0.065447807), (u'n02123159', u'tiger_cat', 0.030858887), (u'n02127052', u'lynx', 0.019239476), (u'n03000247', u'chain_mail', 0.01757087)]]`,
- This PR: `[[(u'n03000247', u'chain_mail', 0.26305375), (u'n03482405', u'hamper', 0.16513243), (u'n02124075', u'Egyptian_cat', 0.057278082), (u'n02346627', u'porcupine', 0.036082812), (u'n02120505', u'grey_fox', 0.035361007)]]`.

Here, changing the parameters (`epsilon=1e-3, momentum=0.999`) of `BatchNormalization` will make the results more closer to the reference: `[[(u'n02124075', u'Egyptian_cat', 0.16711953), (u'n02123045', u'tabby', 0.12701738), (u'n03482405', u'hamper', 0.12486618), (u'n03000247', u'chain_mail', 0.079248495), (u'n02123159', u'tiger_cat', 0.031069454)]]`.

We can set weight values from the reference with the following codes:
```python
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    weights = tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES, scope='module/MobilenetV2')
    values = sess.run(weights)
    values[-2] = np.delete(np.squeeze(values[-2]), 0, axis=-1)
    values[-1] = np.delete(values[-1], 0, axis=-1)

model2.set_weights(values)
```

But, there is still discrepancy even after copying from the values of the reference: `[[(u'n02124075', u'Egyptian_cat', 0.23030871), (u'n02123045', u'tabby', 0.13059281), (u'n03482405', u'hamper', 0.12498391), (u'n03000247', u'chain_mail', 0.044967268), (u'n04209239', u'shower_curtain', 0.023570806)]]`.

I also confirmed that the very first layer (conv+bn+relu6) produced different values, thus I agree with the expectations of @JonathanCMitchell:
> Therefore, it could be the following:
> The way their convolutional kernels operate is different from ours.
> The way their BatchNorm procedure operates is different.
> They are using a different form of image padding.

taehoonlee commented:
Note that the ImageNet performances tested so far were:

|          |  Top-1  |  Top-5  |
|----------|---------|---------|
| baseline | 58.558% | 81.408% |
| bn       | 64.694% | 85.444% |
| bn + hub | 67.058% | 87.156% |

- baseline: the initial version of this PR,
- bn: updating BatchNormalization parameter as `epsilon=1e-3, momentum=0.999`,
- hub: changing different weight values from `tensorflow_hub`.

JonathanCMitchell commented:
@taehoonlee Ok great, I'll get it done today and then I will compare our results to make sure they are the same. Thanks!

taehoonlee commented:
@JonathanCMitchell, TF-hub includes normalization `(x * 2) - 1)` in the graph. In other words, users should put an image ~ [0, 1], then the TF-hub model changes the scale as [-1, 1] and starts feed-forwarding.

JonathanCMitchell commented:
@taehoonlee Ok. I am in the process of rerunning the tests before I push the latest. I just re did all the weights and published them as release [v1.1](https://github.com/JonathanCMitchell/mobilenet_v2_keras/releases/tag/v1.1)

taehoonlee commented:
@JonathanCMitchell, I tested with the default parameters of `tf.contrib.layers` (`epsilon=1e-3, momentum=0.999`). Please note that the TF-hub models use `tf.slim` (see [codes](https://github.com/tensorflow/models/blob/master/research/slim/nets/mobilenet/mobilenet_v2.py)) and the layers in the `tf.slim` have been integrated with `tf.contrib.layers` (see [codes](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/slim/__init__.py)).

JonathanCMitchell commented:
@taehoonlee Please let me know if everything I pushed is acceptable. I will check this later tonight.

taehoonlee commented:
Thank you for the revisions, @JonathanCMitchell 👏! My last requests are:

- Line 54: `[mobilenet_v2_0.35_224)` -> `[mobilenet_v2_0.35_224]`,
- Line 60: `all 16 models` -> `all 22 models`,
- Line 154: `"""` -> `"""Instantiates the MobileNetV2 architecture.` followed by a newline,
- Line 155: `To load a MobileNet` -> `To load a MobileNetV2`
- Line 173: `the MobileNetv2 paper` -> `the MobileNetV2 paper`,
- Line 415-417: Please replace `pool_size = ...; x = AveragePooling2D(pool_size=pool_size)(x); x = Flatten()(x)` with `x = GlobalAveragePooling2D()(x)` in order to enable other backends to work,
- Please clean up the commits, merging appears to be done strangely.

JonathanCMitchell commented:
@taehoonlee Ok I'll get that done tonight, and I'll squash all the commits into one once I finally get everything together. Thanks!

tanakataiki commented:
Thanks for the great work . 
1.Would you tell me When can we(normal user like me) import mobilenetv2 from keras application?
2.Is it possible to upgrade latest keras version here.

bobyguo commented:
@tanakataiki 

> Alternatively:
> 
>  install Keras from the GitHub source:
> First, clone Keras using git:
> 
> git clone https://github.com/keras-team/keras.git
> Then, cd to the Keras folder and run the install command:
> 
> cd keras
> sudo python setup.py install

taehoonlee commented:
@tanakataiki, You can use the bleeding edge by: `pip install git+https://github.com/keras-team/keras.git`

bonlime commented:
@JonathanCMitchell Thank you for your MobileNetv2 implementation! 
Just used it to make a backbone for my Deeplabv3+ model !

bonlime commented:
Also added small fixes to your code
#10163

JonathanCMitchell commented:
You're welcome @bonlime. I am not sure if they will accept the changes but can you remove the `dropout` docstring and parameter from MobileNetV2 please? I forgot to do that. Thanks. 

Since this network is going to be used as a backbone, it may be harder to isolate its layers with a regular expression in your proposed naming system. That's why I included the `mobl` prefix in naming. If you simply match all the layers that contain `mobl` as a prefix, you can perform operations on them.

I'll leave it to @taehoonlee if he wants to accept it or not.

taehoonlee commented:
@JonathanCMitchell, I think that regular expressions can select any layers users want even if @bonlime's naming are applied. Thus #10163 LGTM.

JonathanCMitchell commented:
@taehoonlee @bonlime Great! Sounds good

nicolefinnie commented:
@JonathanCMitchell  I took a look at the config and didn't see any kernel regularizer (or depthwise/pointwise regularizers) being set in any layer, however, in the original paper, they used the weight decay `0.0004`. Did you implement an equivalent regularization somewhere that I overlooked? Thanks for your great contribution.

bonlime commented:
@nicolefinnie It's a common practice not to add explicit regularizers to Keras models. You can always easily add them by iterating over all layers and adding `kernel_regularizer`.

nicolefinnie commented:
> @nicolefinnie It's a common practice not to add explicit regularizers to Keras models. You can always easily add them by iterating over all layers and adding `kernel_regularizer`.

@bonlime  I see, thank you very much.
